### PR TITLE
[addons] Add targetversion attribute to import element for dependencies in addon.xml

### PR DESCRIPTION
**Pull request by mediaminister**
## Description

This pull request adds support for Py2/3 compatibility in **addon.xml** by adding  a new optional `targetversion` xml attribute to the import element.
You can call this `targetversion` attribute a "**forward compatibility version**".

For instance, define Py2/Py3 compatibility from Kodi 17 to Kodi 19:
```
 <import addon="xbmc.python" version="2.25.0" targetversion="3.0.0"/>
```

## Motivation and Context

After https://github.com/xbmc/xbmc/pull/17188 has been merged, add-on developers maintaining a single Py2/Py3 compatible codebase are forced to submit their add-on **twice** to the official Kodi repository. One time to support older Kodi versions and one time to support the new Kodi 19 Matrix. This is needed because  currently **addon.xml**  doesn't support an xml tag to identify Py2/Py3 compatible add-ons.

With this change all existing functionality is preserved and add-on developers just have to submit **once** to the official Kodi repo to get their add-on published for all compatible Kodi versions.

## How Has This Been Tested?

This has been tested with InputStream Helper add-on on Kodi 19 master and Kodi 18.6 on Linux x64.

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by lrusak** [comment]
Thanks for taking the time to do this but I'm not sure I really understand this solution. You are defining a version you require but also a target? What does the target mean? You would also be able to do this for any required add-on and not just xbmc.python (not sure what the implications are there).

What happens if we have to bump to xbmc.python `3.0.1` and all the add-on devs have to submit their updates again because now all the add-ons will be marked as broken?

**Comment by mediaminister** [comment]
> You are defining a version you require but also a target?

The principle is that older Kodi versions don't support the `target` attribute and always use the `version` attribute. You can define `version="2.25.0"` so your add-on is compatible with Kodi 17 and Kodi 18. 
The `target` attribute is only supported in Kodi 19 Matrix, so if you define  `<import addon="xbmc.python" version="2.25.0" target="3.0.0"/>` Kodi Matrix reads the `target` attribute and ignores the `version` attribute.

> What does the target mean?

It's just a second `version` attribute that only exists in Kodi 19 Matrix, so a single **addon.xml** can support both Python 2 and Python 3 Kodi versions.

> You would also be able to do this for any required add-on and not just xbmc.python (not sure what the implications are there).

Internally the `target` attribute is treated as `version` attribute. So, it just "overwrites" the `version` attribute if it is defined.

> What happens if we have to bump to xbmc.python 3.0.1 and all the add-on devs have to submit their updates again because now all the add-ons will be marked as broken?

This change doesn't change the internal logic, `target` "overwrites" the `version` attribute if it is defined.

**Comment by ronie** [comment]
hmm, i don't think this logic is future proof?

let's say we bump xbmc.python to 4.0.0 in Kodi 20.

an addon that is compatible with kodi 18 / kodi 19 / kodi 20 would then need to use
`<import addon="xbmc.python" version="2.25.0" target="4.0.0"/>`

as far as i can see, the result would be the addon works with kodi 18 & kodi 20, but will be marked as broken on kodi 19.

**Comment by mediaminister** [comment]
> hmm, i don't think this logic is future proof?

Not yet, I should read the abi version and compare this with the target version.

**Comment by CastagnaIT** [comment]
in alternative you can also try add the controlling characters
as i tried to explain in my post: https://github.com/xbmc/xbmc/issues/17326#issuecomment-583607750

**Comment by knaerzche** [comment]
I would rather suggeset implement a "maxversion" attribute, since "targetversion" should be represented by "version" already.
With minversion attribute being already there, this would then implement all logic as suggested by @CastagnaIT, and would be easier to implement as well: From what I found it just would have to be added to this condition: https://github.com/xbmc/xbmc/blob/master/xbmc/addons/addoninfo/AddonInfo.cpp#L180-L183. 
@lrusak is right: it also has to be taken care what implication this has when importing other addons than xbmc.python. This maxversion attribute would help addons, where the latest version of an addon dependency breaks their own functionality, to make clear that it is not compatible with this latest version.

**Comment by mediaminister** [comment]
> I would rather suggeset implement a "maxversion" attribute, since "targetversion" should be represented by "version" already.

I implemented `maxversion` now.

**Comment by knaerzche** [comment]
@mediaminister: Thanks for already eleminitating the shortcommings of my suggestens by comparing to m_minversion - I guess logic should be alright like that (provided that, unlike previously, this https://github.com/xbmc/xbmc/blob/master/addons/xbmc.python/addon.xml#L3 is now changed regularly). without wanting to pre-empt the codereview now - I think  this first `versionMax.empty()` at https://github.com/xbmc/xbmc/pull/17435/commits/c08b4edd77c94da7a2a9e3a33d24952e28932a40#diff-a91dcb9aea88dd59d0672060a07419f2R183 should be 
`version.empty()`

**Comment by romanvm** [comment]
I haven't been following this discussion recently because of various personal stuff, but I have another solution in mind: Python3 Matrix+ addon will have `xbmc.python` version equal or grater than `3.0.0`, while Python 3 compatible addons for lower Kodi versions should have `<python3 />` tag in their metadata. And in Matrix any addon with `xbmc.python` versions lower than `3.0.0` and without `<python3 />` tag will be marked as broken and incompatible. This may eliminate the need in double submission of addons to matrix and lower branches.

If this solution looks good, please give me about a week to make a proof of concept.

**Comment by mediaminister** [comment]
@romanvm 

> `<python3 />` tag

Such a tag is not future proof. For instance, when Python 4 will be needed.

> This may eliminate the need in double submission of addons to matrix and lower branches.

This pull request also ensure a single submission workflow, adding the following line to **addon.xml** and submitting to krypton branch should make you add-on available for krypton, leia and matrix.

` <import addon="xbmc.python" version="2.25.0" maxversion="3.0.0"/>`

**Comment by knaerzche** [comment]
I'd prefer this current soultion, since it's more general and can be used by none "xbmc.python" addon imports as well.

**Comment by knaerzche** [comment]
@mediaminister Maybe you should ajust the subject of this PR - it was originally initiated by this PY2/PY3 issue - but its now going beyond this, by introducing this new "maxversion" attribute which is relevant to all addons. "xbmc.python" is just a usecase for this. 
I guess this will help when looking at this PR in years of time.

File: xbmc/addons/addoninfo/AddonInfo.cpp [hunk 1]
```diff
@@ -177,9 +177,11 @@ bool CAddonInfo::ProvidesSeveralSubContents() const
   return contents > 0 ? true : false;
 }
 
-bool CAddonInfo::MeetsVersion(const AddonVersion& versionMin, const AddonVersion& version) const
+bool CAddonInfo::MeetsVersion(const AddonVersion& versionMin, const AddonVersion& version,
+                              const AddonVersion& versionMax) const
 {
-  return !(versionMin > m_version || version < m_minversion);
+  return !(versionMin > m_version || (versionMax.empty() && version < m_minversion) ||
```
Comment (enen92) [inline xbmc/addons/addoninfo/AddonInfo.cpp]: Using the negation of the overall condition makes this tremendous hard to read (and error-prone) as the number of subconditions increase. Can you write it the other way around?

**Comment by enen92** [comment]
Overall I like the approach. I'm thinking this might cause interpretation problems to addons such as script.modules.

Suppose that my addon depends on something like `script.module.foo`. It is proven to work for any version of the module comprised between 1.0.0 and 1.5.0. However it breaks for version 2.0.0.
As a dev I would define that dependency as a range in my addon addon.xml:

`<import addon="script.module.foo" version="1.0.0" maxversion="1.5.0"/>`

Regarding `script.module.foo`, it is py2 and py3 compatible since version 1.0.0. So, its addon xml:

1.0.0: `<import addon="xbmc.python" version="2.26.0" maxversion="3.0.0"/>`
1.5.0: `<import addon="xbmc.python" version="2.26.0" maxversion="3.0.0"/>`
2.0.0: `<import addon="xbmc.python" version="2.26.0" maxversion="3.0.0"/>`

I am the one that initially required `script.module.foo` to be in the repository, so I was the one who first submitted it (versions 1.0.0 and version 1.5.0). Later on someone requires `script.module.foo` version 2.0.0 and submits it to the repo. The repository generator updates the repository, packages 2.0.0 and that is the version everyone will receive from now on. My addon just gets disabled because of a dependency...

**Comment by knaerzche** [comment]
> My addon just gets disabled because of a dependency...
But isn't that the way dependency mangement is supposed to work?

For this particular case:
If an addon sets this optional maxversion attribute in its addon.xml implies, that its noticed that functionality breaks at a certain point (which can only be the case after script.module.foo > 1.5.0 was released). This means that there have to be workarounds in that addon itself, which check the version of script.module.foo before using its functionality. So: the addon does not really depend for  script.module.foo to be available in a useable version and should therefore defined as  

`<import addon="script.module.foo" version="1.0.0" maxversion="1.5.0" optional="true"/>`

**Comment by enen92** [comment]
@knaerzche it strongly depends on the way you read "optional". For me optional is something that you may not define but that should work if you define it. 

In the situation exemplified above if `maxversion` is not defined the worst thing that may happen is that my addon suddenly starts throwing logic errors and I have to send an update to fix the behaviour. If `maxversion` is defined it will simply get automatically disabled for all users and I can't do anything about it. I know it worked before and kodi version itself hasn't changed (plus my 1.5 version is on the mirrors) so I expect to just be able to use my addon.

**Comment by knaerzche** [comment]
> @knaerzche it strongly depends on the way you read "optional". For me optional is something that you may not define but that should work if you define it.

For me in this case this mainly meant, that the addon can still be installed, even if no script.module.foo >= 1.0.0 and <= 1.5.0 is available ... which should be technically correct according to https://kodi.wiki/view/Addon.xml.
Since setting this "maxversion" attribute would mean a resubmission of that addon, I assumed that the developper either makes sure that that his addon works with versions > 1.5.0 (than no maxversion would  be necessary at all) or builds workarounds in his addon that some kind of handle this discrepancies. Thats at least how would handle that - but it might be foolish to assume everybody would
So: you're right: It obviously leads to interpreation issues and should be documented well.

**Comment by phunkyfish** [comment]
It sounds to me like multiple versions of the script module's need to be supported in the repo. Otherwise script module updates can break many addons.

File: xbmc/addons/addoninfo/AddonInfo.cpp (hunk 1 above)
Comment (mediaminister) [reply xbmc/addons/addoninfo/AddonInfo.cpp]: Okay, I rewrote this and made sure that `versionMax` is never empty by using the `version` value:
https://github.com/xbmc/xbmc/blob/3fb7ad61c9f6451ba1bc0283d013230bafb2c2f6/xbmc/addons/addoninfo/AddonInfo.h#L39

So the overall condition could be simplified to:
https://github.com/xbmc/xbmc/blob/3fb7ad61c9f6451ba1bc0283d013230bafb2c2f6/xbmc/addons/addoninfo/AddonInfo.cpp#L183

**Comment by mediaminister** [comment]
@enen92 @phunkyfish 

> My addon just gets disabled because of a dependency...

That might be possible in theory but I tested this and I found out  that only the xbmc.* and kodi.* dependencies are checked:

https://github.com/xbmc/xbmc/blob/a689d1c746cc229aa9ac8c004affc9f6a7fb4594/xbmc/addons/AddonManager.cpp#L713-L732

**Review by AlwinEsch — CHANGES_REQUESTED** [review]
Also I'm not sure if that is the best way, but I'm mainly a binary addon man. Let see what my colleagues say.

File: cmake/scripts/common/AddonHelpers.cmake [hunk 2]
```diff
@@ -126,10 +127,10 @@ macro (build_addon target prefix libs)
                 if(include_name AND CMAKE_MATCH_3 MATCHES ^${depend_header})
                   get_directory_property(CURRENT_DEFS COMPILE_DEFINITIONS)
                   if(NOT used_type_name IN_LIST CURRENT_DEFS)
-                    set(ADDON_DEPENDS "${ADDON_DEPENDS}\n<import addon=\"${${xml_entry_name}}\" minversion=\"${${depends_minver}}\" version=\"${${depends_ver}}\"/>")
+                    set(ADDON_DEPENDS "${ADDON_DEPENDS}\n<import addon=\"${${xml_entry_name}}\" minversion=\"${${depends_minver}}\" version=\"${${depends_ver}}\" maxversion=\"${${depends_maxver}}\"/>")
                     # Inform with them the addon header about used type, if not present before
                     add_definitions(-D${used_type_name})
-                    message(STATUS " - Added API usage definition: ${used_type_name} (Version: \"${${depends_ver}}\", Min. Version: \"${${depends_minver}}\")")
+                    message(STATUS " - Added API usage definition: ${used_type_name} (Version: \"${${depends_ver}}\", Min. Version: \"${${depends_minver}}\", Max. Version: \"${${depends_maxver}}\")")
```
[the remark this answers is not in the document]
Comment (mediaminister) [reply cmake/scripts/common/AddonHelpers.cmake]: Okay, I removed these changes from my pull request.

**Comment by tamland** [comment]
I recommend everyone to take a look at [how android solves this issue](https://developer.android.com/guide/topics/manifest/uses-sdk-element). Usually, specifying a maximum version doesn't make any sense. What you normally want is min + target. (PS: the current `version` attribute in kodi behaves like androids `minSdkVersion`)

**Comment by mediaminister** [comment]
@tamland It's indeed better to call this attribute `targetversion`, as it's intended to give add-on developers the opportunity to declare forward-compatibility. This doesn't change the implementation.

**Comment by phunkyfish** [comment]
As the decision was to go for a hard break I will close this PR and mark as obsolete.

**Comment by knaerzche** [comment]
@mediaminister: As I predicted in https://github.com/xbmc/xbmc/pull/17435 - the work was not worth it because "the decision" was made for nothing other than ideological reasons (even if they said they were said only were looking for someone to implement it). Unfortunately, Kodi is no longer community-focused software with the people who make decisions these days.

**Pull request closed without merging**